src/Common/NamedCollections/NamedCollectionsMetadataStorage.cpp

```diff
@@ -0,0 +1,491 @@
+
+        out.next();
+        if (getContext()->getSettingsRef().fsync_metadata)
+            out.sync();
+        out.close();
+
+        fs::rename(tmp_path, getPath(path));
+    }
+
+    void remove(const std::string & path) override
+    {
+        if (!removeIfExists(getPath(path)))
+        {
+            throw Exception(
+                ErrorCodes::NAMED_COLLECTION_DOESNT_EXIST,
+                "Cannot remove `{}`, because it doesn't exist", path);
+        }
+    }
+
+    bool removeIfExists(const std::string & path) override
+    {
+        return fs::remove(getPath(path));
+    }
+
+private:
+    std::string getPath(const std::string & path) const
+    {
+        return fs::path(root_path) / path;
+    }
+
+    /// Delete .tmp files. They could be left undeleted in case of
+    /// some exception or abrupt server restart.
+    void cleanup()
+    {
+        std::vector<std::string> files_to_remove;
+        for (fs::directory_iterator it{root_path}; it != fs::directory_iterator{}; ++it)
+        {
+            const auto & current_path = it->path();
+            if (current_path.extension() == ".tmp")
+                files_to_remove.push_back(current_path);
+        }
+        for (const auto & file : files_to_remove)
+            fs::remove(file);
+    }
+};
+
+
+class NamedCollectionsMetadataStorage::ZooKeeperStorage : public INamedCollectionsStorage, private WithContext
+{
+private:
+    std::string root_path;
+    mutable zkutil::ZooKeeperPtr zookeeper_client{nullptr};
+    mutable zkutil::EventPtr wait_event;
+
+public:
+    ZooKeeperStorage(ContextPtr context_, const std::string & path_)
+        : WithContext(context_)
+        , root_path(path_)
+    {
+        if (root_path.empty())
+            throw Exception(ErrorCodes::INVALID_CONFIG_PARAMETER, "Collections path cannot be empty");
+
+        if (root_path != "/" && root_path.back() == '/')
+            root_path.resize(root_path.size() - 1);
+        if (root_path.front() != '/')
+            root_path = "/" + root_path;
+
+        auto client = getClient();
+        if (root_path != "/" && !client->exists(root_path))
+        {
+            client->createAncestors(root_path);
+            client->createIfNotExists(root_path, "");
+        }
+    }
+
+    ~ZooKeeperStorage() override = default;
+
+    bool supportsPeriodicUpdate() const override { return true; }
+
+    void waitUpdate(size_t timeout) override
+    {
+        if (wait_event)
+            wait_event->tryWait(timeout);
+    }
+
+    std::vector<std::string> list() const override
+    {
+        if (!wait_event)
+            wait_event = std::make_shared<Poco::Event>();
+        return getClient()->getChildren(root_path, nullptr, wait_event);
+    }
+
+    bool exists(const std::string & path) const override
+    {
+        return getClient()->exists(getPath(path));
+    }
+
+    std::string read(const std::string & path) const override
+    {
+        return getClient()->get(getPath(path));
```

Comment:
we should also think about caching, depending on how often we read named collections
but it can be done as future imporovement